lib/charms/ip_router_interface/v0/ip_router_interface.py

```diff
@@ -27,6 +130,274 @@
+            A list of objects of type `Network`. This list contains networks
+            from all ip-router providers that are integrated with the charm.
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        router_relations = self.model.relations.get("ip-router")
+        all_networks = []
+        for relation in router_relations:
+            if networks := relation.data[relation.app].get("networks"):
+                all_networks.extend(json.loads(networks))
+        return all_networks
+
+    def _validate_network(self, network_request: Network, new_networks: List[Network]) -> bool:
+        """Validates the network configuration created by the ip-router requirer
+
+        The requested network must have all of the required fields, the gateway
+        has to be located within the network, and all of the routes need to have
+        a path through the top level network. The requested network must also be
+        unassigned by the provider.
+
+        Args:
+            network_request:
+                An object of type `Network` that will be validated.
+            new_networks:
+                The rest of the networks to check if this one is mutually
+                exclusive to the rest of the subnets.
+
+        Raises:
+            ValueError:
+                Reasons could be that the gateway is not within the network,
+                there is no route to the destination, the network is already
+                taken or the same network is requested twice.
+            KeyError:
+                Missing required key
+        """
+        if "gateway" not in network_request:
+            raise KeyError("Key 'gateway' not found.")
+
+        if "network" not in network_request:
+            raise KeyError("Key 'network' not found.")
+
+        gateway = IPv4Address(network_request.get("gateway"))
+        network = IPv4Network(network_request.get("network"))
+
+        if gateway not in network:
+            ValueError("Chosen gateway not within given network.")
+
+        for route in network_request.get("routes", []):
+            if "gateway" not in route:
+                raise KeyError("Key 'gateway' not found in route.")
+
+            if "destination" not in route:
+                raise KeyError("Key 'destination' not found in route.")
+            route_gateway = IPv4Address(route["gateway"])
+            if route_gateway not in network:
+                raise ValueError("There is no route to this destination from the network.")
+
+        for entry in new_networks:
+            if entry == network_request:
+                continue
+            existing_network = IPv4Network(entry["network"])
+            new_network = IPv4Network(network_request["network"])
+
+            if new_network.subnet_of(existing_network) or new_network.supernet_of(
+                existing_network
+            ):
+                raise ValueError("This network has been defined in another entry.")
```

Comment:
This code block is repeated and suggesting to define it as a separate method that is called when needed.

Reply:
Removed the repetition and combined the two different lists.